### PR TITLE
EPUB: more support for RTL documents

### PR DESCRIPTION
**Pull request by poire-z**
bump crengine: enhanced RTL block rendering. Includes:
- Re-enable: Use the same FT load flags in HB as in FT https://github.com/koreader/crengine/pull/311
- Avoid redundant calls to getStyle() https://github.com/koreader/crengine/pull/312
- CSS: parse and store "direction:"
- Enhanced block rendering: handle RTL direction
- Enhanced block rendering: handle RTL tables
- Pagesplitting: account for lines' text direction
- CSS: add support for more pseudo-classes
- epub.css: a few updates for RTL documents
- Fix LVDocView::getBookmark() which could be slow or wrong https://github.com/koreader/crengine/pull/313

Add a few style tweaks useful to RTL readers.

Wikipedia Save as EPUB: build proper RTL documents from persian, arabic, hebrew (and others) Wikipedia articles.

Screenshots in https://github.com/koreader/crengine/pull/312

### CODE REVIEW

File: frontend/ui/data/css_tweaks.lua [hunk 1]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
-                priority = 3, -- so it overrides the ones above
+                priority = 4, -- so it overrides the ones above
+            },
+        },
+        {
+            title = _("Text direction"),
+            {
+                title = _("About text direction"),
+                info_text = _([[
+Documents in languages like arabic or hebrew are to be rendered Right-To-Left. This does not only affects text layout, but also list items bullets and numbers (put on the right) and table layout (cells laid from right to left).
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]: Maybe something like this? (In any case, Arabic and Hebrew are capitalized in English. ;-))
```suggestion
Languages like Arabic or Hebrew use right-to-left writing systems (Right-To-Left or RTL). This doesn't only affect text layout, but also list item bullets and numbers, which have to be put on the right side of the page, as well as and tables, where the cells are laid out from right to left.
```

File: frontend/ui/data/css_tweaks.lua [hunk 2]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
+
+Usually, the publisher has set the appropriate tags to force such rendering - but if he hasn't, or if you're reading plain text documents, you may want to manually enable it with these tweaks.
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
Usually, the publisher will have set the appropriate tags to enable RTL rendering. But if these are missing, or if you're reading plain text documents, you may want to manually enable RTL with these tweaks.
```

File: frontend/ui/data/css_tweaks.lua [hunk 3]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
+Note that in the absence of such specifications KOReader will try to detect the language of each paragraph and set the approppriate rendering per paragraph.
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
Note that in the absence of such specifications, KOReader will try to detect the language of each paragraph to set the appropriate rendering per paragraph.
```

File: frontend/ui/data/css_tweaks.lua [hunk 4]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
+
+You may also want to enable, in top menu → Gear → Navigation →, Invert page turn taps and swipes.]]),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
You may also want to enable, in the top menu → Gear → Navigation →, Invert page turn taps and swipes.]]),
```

File: frontend/ui/data/css_tweaks.lua [hunk 5]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
+                separator = true,
+            },
+            {
+                id = "body_direction_rtl",
+                title = _("Set document direction to RTL (Right-to-Left)"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]: Seems a bit wordy. Couldn't we suffice with something like this?
```suggestion
                title = _("Force RTL (Right-to-Left)"),
```

Or even (given the explanation item above):

```suggestion
                title = _("Force RTL"),
```

File: frontend/ui/data/css_tweaks.lua [hunk 6]
```diff
@@ -161,7 +161,48 @@ h1 + h6, h2 + h6, h3 + h6, h4 + h6, h5 + h6 { page-break-before: avoid !importan
...
+                id = "text_align_all_right",
+                title = _("Right align all elements"),
+                description = _("Enforce right alignment of text in all elements."),
+                css = [[* { text-align: right !important; }]],
+                priority = 3, -- so it overrides the ones from Text alignment
+                separator = true,
+            },
+            {
+                id = "body_direction_ltr",
+                title = _("Set document direction to LTR (Left-to-Right)"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]: Same comment as above.

File: frontend/ui/wikipedia.lua [hunk 7]
```diff
@@ -655,6 +655,31 @@ local ext_to_mimetype = {
     woff = "application/font-woff",
 }
 
+-- Display from these Wikipedia should be RTL
```
Comment (Frenzie) [inline frontend/ui/wikipedia.lua]: Don't be afraid to use multiline comments. ;-)

File: frontend/ui/wikipedia.lua [hunk 8]
```diff
@@ -916,19 +941,19 @@ h1, h2 {
...
     page-break-inside: avoid;
     page-break-after: avoid;
-    text-align: left;
+    text-align: start;
 }
 /* Styles for our centered titles on first page */
 h1.koreaderwikifrontpage, h5.koreaderwikifrontpage {
     page-break-before: avoid;
     text-align: center;
-    margin-top: 0em;
```
Comment (Frenzie) [inline frontend/ui/wikipedia.lua]: Huh, I somehow overlooked these needless units?

File: frontend/ui/data/css_tweaks.lua (hunk 5 above)
Comment (poire-z) [reply frontend/ui/data/css_tweaks.lua]: Well, these tweaks do not "force" RTL or LTR. They just set it to the top node. It will be overriden if there are dir= below in the document.
I have added no tweak that "force" a direction (I guess they would need `* {direction: rtl !important;}` - and I don't think they would be useful).

Also, after all this, I don't quite like the wording "text direction", because, as we see, the text itself does not change direction (english stays Left-To-Right). "Document direction" is a bit less confusing, although it could give the idea it means "invert page turn taps and swipes" - that's why I added that little help text.
So, I don't know :) other suggestions ?

File: frontend/ui/wikipedia.lua (hunk 7 above)
Comment (poire-z) [reply frontend/ui/wikipedia.lua]: Remind me what are these and how they differ from consecutive `-- blah` ?! :)

File: frontend/ui/data/css_tweaks.lua (hunk 5 above)
Comment (Frenzie) [reply frontend/ui/data/css_tweaks.lua]: How about just "RTL"? Or `Document direction RTL` (it doesn't need to be a sentence).

File: frontend/ui/wikipedia.lua (hunk 7 above)
Comment (Frenzie) [reply frontend/ui/wikipedia.lua]: I just mean 
```
--[[--lua
Bah blah.

Blah.
--]]
```

**Comment by poire-z** [comment]
Small (or not so small) issue noticed on the crengine side.
Because of [this change](https://github.com/koreader/crengine/commit/5967f723b8937fcac30739732d61db8d7af0eee7#diff-77f6525e33d2fd1aeeb073a64dd8d759R13444-R13460) to get selection working in table cells, the getBookmark() call used to get the last_xpointer for each page can, on some pages (the pages with notes on that Petra.AR.epub), be slow (1-2 second lag in page turn when on some pages on my Kobo), or be wrong (same pages when in scroll mode, we'd always get back to the page with note 1 when we were on some other page from that list than spans 4 or 5 pages).
So, let's wait for another crengine bump before merging this...

File: frontend/ui/wikipedia.lua (hunk 7 above)
Comment (poire-z) [reply frontend/ui/wikipedia.lua]: Well, I find these uglier :)

File: frontend/ui/data/css_tweaks.lua (hunk 5 above)
Comment (poire-z) [reply frontend/ui/data/css_tweaks.lua]: That's fine :) Done.
